### ISSUE TITLE
IIOProviderContextListener -> NoClassDefFoundError with apache pdfbox JBIG2 ImageIO plugin

### ISSUE THREAD

**Issue by Zardoz89** [haraldk/TwelveMonkeys#503]
I don't sure if the problem is with IIOProviderContextListener or in Apache PDFBox JBIG2 ImageIO plugin, but sometimes I get a NoClassDefFoundError when I restart a servlet aplication on a Tomcat.

```
java.lang.NoClassDefFoundError: Could not initialize class org.apache.pdfbox.jbig2.JBIG2ImageReader
    at org.apache.pdfbox.jbig2.JBIG2ImageReaderSpi.createReaderInstance(JBIG2ImageReaderSpi.java:117)
    at javax.imageio.spi.ImageReaderSpi.createReaderInstance(ImageReaderSpi.java:320)
    at javax.imageio.ImageIO$ImageReaderIterator.next(ImageIO.java:529)
    at javax.imageio.ImageIO$ImageReaderIterator.next(ImageIO.java:513)
    at org.apache.pdfbox.filter.Filter.findImageReader(Filter.java:155)
    at org.apache.pdfbox.filter.JBIG2Filter.decode(JBIG2Filter.java:67)
    at org.apache.pdfbox.cos.COSInputStream.create(COSInputStream.java:84)
    at org.apache.pdfbox.cos.COSStream.createInputStream(COSStream.java:175)
    at org.apache.pdfbox.pdmodel.common.PDStream.createInputStream(PDStream.java:241)
    at org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject.createInputStream(PDImageXObject.java:735)
    at org.apache.pdfbox.pdmodel.graphics.image.SampledImageReader.from1Bit(SampledImageReader.java:253)
    at org.apache.pdfbox.pdmodel.graphics.image.SampledImageReader.getRGBImage(SampledImageReader.java:212)
    at org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject.getImage(PDImageXObject.java:444)
    at org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject.getImage(PDImageXObject.java:425)
    at org.apache.pdfbox.rendering.PageDrawer.drawImage(PageDrawer.java:1100)
    at org.apache.pdfbox.contentstream.operator.graphics.DrawObject.process(DrawObject.java:63)
    at org.apache.pdfbox.contentstream.PDFStreamEngine.processOperator(PDFStreamEngine.java:869)
    at org.apache.pdfbox.contentstream.PDFStreamEngine.processStreamOperators(PDFStreamEngine.java:505)
    at org.apache.pdfbox.contentstream.PDFStreamEngine.processStream(PDFStreamEngine.java:479)
    at org.apache.pdfbox.contentstream.PDFStreamEngine.processPage(PDFStreamEngine.java:152)
    at org.apache.pdfbox.rendering.PageDrawer.drawPage(PageDrawer.java:263)
    at org.apache.pdfbox.rendering.PDFRenderer.renderImage(PDFRenderer.java:321)
    at org.apache.pdfbox.rendering.PDFRenderer.renderImage(PDFRenderer.java:243)
    at org.apache.pdfbox.rendering.PDFRenderer.renderImage(PDFRenderer.java:203)
    at com.digibis.digiimage.pdf.pdfbox.PdfBoxPdfProcessor.getImageFromDocumentPage(PdfBoxPdfProcessor.java:105)
 ...
```

**Comment by haraldk** [comment]
Hi Luis,

The problem can be either of those, or something else. It's hard to say by just looking at the stack trace. 

The `JBIG2ImageReaderSpi` seems simple enough, so it shouldn't cause problems. My educated guess is that you are missing some (or have bad versions of) some of the support classes needed by the `JBIG2ImageReader` (because of the "Could not initialize class" message). Perhaps you have multiple web apps? Or have multiple sources for the `JBIG2ImageReaderSpi` class?

I'd use a debugger, to see what's happening. Do you have multiple versions of `JBIG2ImageReaderSpi` in the `IIORegistry` when this happens? Is the class in the registry loaded by the current apps class loader? 

Unfortunately, there's not much more help I can give you in this case.

-- 
Harald K

**Comment by rich-coe** [comment]
I found the following piece of code helpful for debugging NoClassDef errors
Create a manifest file with the required Premain-Class
Create the jar tracer.jar
Load the jar with -javaagent option on your app startup and direct standard out to a file.
Search for when the failing class is loaded, and the messages will show which class cannot be found.


    import java.lang.instrument.*;
    import java.security.*;

    // manifest.mf
    // Premain-Class: ClassLoadTracer

    // jar -cvfm tracer.jar manifest.mf ClassLoaderTracer.class

    // java -javaagent:tracer.jar  [...]

    public class ClassLoadTracer 
    {
        public static void premain(String agentArgs, Instrumentation inst) 
        {
            final java.io.PrintStream out = System.out;
            inst.addTransformer(new ClassFileTransformer() {
                public byte[] transform(ClassLoader loader, String className, Class classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {

                    String pd = (null == protectionDomain) ? "null" : protectionDomain.getCodeSource().toString();
                    out.println(className + " loaded by " + loader + " at " + new java.util.Date() + " in " + pd);

                    // dump stack trace of the thread loading class 
                    Thread.dumpStack();

                    // we just want the original .class bytes to be loaded!
                    // we are not instrumenting it...
                    return null;
                }
            });
        }
    }

**Comment by haraldk** [comment]
Related, from the [PDFBox mailing list](http://mail-archives.apache.org/mod_mbox/pdfbox-users/201808.mbox/%3CCAAPHLv_fXGmB18pobaNEnus0EMQLuqD-gBb5a_y8QrWr-HizEg@mail.gmail.com%3E). Unfortunately, there doesn't seem to be a solution there. 

-- 
Harald K

**Comment by Zardoz89** [comment]
I raised an issue on PDFBox jira : https://issues.apache.org/jira/browse/PDFBOX-4671

The most strange thing is that this problem happens randomly.

> Perhaps you have multiple web apps? Or have multiple sources for the JBIG2ImageReaderSpi class?

Yes, we have multiples instances on the same tomcat. It's a develop server where we have a few products (all based over the same core product that have the same ImageIO plugins). And when this happens, looks that happens on all web applications at same time.

**Comment by Zardoz89** [comment]
I prepared a minimal servlet application that can reproduce this fail case, if someone start/stop from the Tomcat manager enough times two instances.

[jbig2 servlet failcase.tar.gz](https://github.com/haraldk/TwelveMonkeys/files/3742956/jbig2.servlet.failcase.tar.gz)

**Comment by haraldk** [comment]
> I see that fails when tries to create a instance of JBIG2ImageReader class. This is weird... this class is on the same package that JBIG2ImageReaderSpi

No, not really. The problem is (most likely) that the Spi class is an "old" instance (from a previous web app), and thus has a different class loader, than the reader implementation. Two otherwise completely equal classes (ie. same package name, class name, perhaps even originating from the same file on disk) with different class loaders, are treated as completely different and unrelated by the JVM. 
You should be able to see this (ie. what class loader the classes have) in a debugger, and this *might* help you locate the problem. What makes this hard to pinpoint, is that the "error" happens silently a long time *before* the symptoms show up. 

It's enough if one of the web apps in the Tomcat instance does not use the context listener, or if one of  if the apps (or any other library or even Tomcat itself) invokes `scanForPlugins()`, either before the context listener does, or with a different class loader, and this will happen.  


Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
PS: I just discovered from the PDFBox issue you mention that you are 

> additionally doing a `ImageIO.scanForPlugins()`

This may actually be the cause of the problem in itself, and not a step towards the solution (although, in an ideal world, this invocation should not discover any new plugins, and simply be a no-op).

-- 
Harald K

**Comment by Zardoz89** [comment]
> This may actually be the cause of the problem in itself, and not a step towards the solution (although, in an ideal world, this invocation should not discover any new plugins, and simply be a no-op).

Yeah.... I remove these line and relaunch my small test case. Keeps happening.

**Comment by Zardoz89** [comment]
**EUREKA!** 
It's these two services! It's the exactly the same problem that your listener fixes for the ImageIO SPI service, but it happens with his CacheFactory and LoggerFactory services.
I removed both and replaced his Logger wraper to JUL, by SLF4J and his CacheFactory by a simple instance of his cache implementation. Now  I can't reproduce anymore the bug!
So, or they does what I did or I/they create a listener to unregister his services like you Listener does for ImageIO SPI
